### PR TITLE
feat: enhance OTel token and HTTP instrumentation

### PR DESCRIPTION
**Pull request by doughayden**
## What

Backport OTel enhancements from kairos: opt-in FastAPI HTTP-layer instrumentation, canonical `gen_ai.usage.*` semconv attributes for cached / reasoning / tool-use tokens, and a new Cloud Run concurrency tuning reference doc.

## Why

- The genai instrumentor already auto-emits `input_tokens` / `output_tokens`, but cached, reasoning (`thoughts_token_count`), and tool-use prompt tokens were either missing or recorded under non-canonical `gen_ai.usage.experimental.*` names. Aligning with OTel semconv (`cache_read.input_tokens`, `reasoning_tokens`, `tool_use.input_tokens`) makes dashboards portable and removes the `experimental.` prefix that signaled instability.
- `total_tokens` duplicated information derivable from the component counts and has no semconv equivalent — dropping it trims log noise and avoids publishing a non-standard span attribute.
- FastAPI HTTP-layer spans were previously undocumented-but-commented-out. Making it a first-class opt-in via `setup_opentelemetry(app=...)` lets downstream projects with custom routes add server spans without forking observability setup.
- The `ValueError` suppression on `otel_context.detach()` is unnecessary once `attach` and `detach` both run inside the background task's own context — replacing the suppression with an invariant comment clarifies intent.
- Cloud Run concurrency tuning is the most commonly-asked scaling question for downstream consumers; a reference doc captures the runtime model, memory math, GIL / process tradeoffs, and starting-point profile in one place.

## How

**Code**

- `utils/observability.py` — add optional `app: FastAPI | None = None` parameter; when provided, call `FastAPIInstrumentor().instrument_app(app)` after tracer provider setup; reorder instrumentors to run after tracer/logger provider registration
- `callbacks.py` — replace `total_tokens` / `experimental.cached_tokens` / `experimental.total_tokens` with canonical `gen_ai.usage.cache_read.input_tokens`, `gen_ai.usage.reasoning_tokens`, `gen_ai.usage.tool_use.input_tokens`; update log dict keys to match (`cached_tokens`, `reasoning_tokens`, `tool_use_tokens`)
- `pyproject.toml` / `uv.lock` — add `opentelemetry-instrumentation-fastapi>=0.58b0,<0.59`

**Tests**

- `conftest.py` — expand `MockUsageMetadata` with `thoughts_token_count` / `tool_use_prompt_token_count`; add `mock_span` fixture that patches `trace.get_current_span`
- `test_logging_callbacks.py` — add log + span assertions for the two new attributes; refactor existing span test to use `mock_span`; update negative assertions for the removed `total_tokens` key

**Docs**

- `docs/observability.md` — new "Custom Gen AI Usage Attributes" table
- `docs/references/opentelemetry-architecture.md` — rework FastAPI section (opt-in activation, `instrument_app` vs `.instrument()` rationale, dual `http send` span note); drop stale `ValueError` suppression rationale
- `docs/references/cloud-run-concurrency-tuning.md` — new deep-dive; linked from `docs/references/README.md`
- `docs/references/testing.md` — add "Test Double Naming" table
- `AGENTS.md` — asyncpg type strictness note; test-double naming convention

## Breaking Changes

- `LoggingCallbacks.after_model` span attributes renamed:
  - `gen_ai.usage.experimental.cached_tokens` → `gen_ai.usage.cache_read.input_tokens`
  - `gen_ai.usage.experimental.total_tokens` → **removed** (derive as `input_tokens + output_tokens + reasoning_tokens + tool_use.input_tokens`)
- Log entry key `total_tokens` removed from the `Token usage:` log line.
- Dashboards or alert rules referencing the `experimental.*` names must be updated before merge rolls to prod.

## Tests

- [ ] `uv run ruff format && uv run ruff check --fix && uv run mypy && uv run pytest --cov` passes locally
- [ ] New unit tests in `test_logging_callbacks.py` cover log output and span attributes for cached / reasoning / tool-use tokens
- [ ] `mock_span` fixture consumed by refactored span assertion tests — no regressions in negative-path test (`test_after_model_skips_span_attributes_when_none`)
- [ ] Manual: deploy locally via `docker compose up --build --watch`, exercise an agent flow that emits `thoughts_token_count` / `tool_use_prompt_token_count`, inspect Cloud Trace for the three new span attributes
- [ ] Manual: call `setup_opentelemetry(..., app=app)` from `server.py` (or a dev harness) and confirm `POST /run` FastAPI server spans appear as parents of ADK `invocation` spans
- [ ] Docs render correctly on GitHub (tables, callouts, internal links in `cloud-run-concurrency-tuning.md`)

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -438,6 +447,14 @@ def _factory(**kwargs: Any) -> MockUsageMetadata:
     return _factory
 
 
+@pytest.fixture
+def mock_span(mocker: MockerFixture) -> MockType:
+    """Mock span returned by ``trace.get_current_span()`` in callbacks."""
+    span = mocker.Mock()
+    mocker.patch("agent_foundation.callbacks.trace.get_current_span", return_value=span)
```
[the remark this answers is not in the document]
Comment (doughayden) [reply tests/conftest.py:454]: false positive, ci already passed

File: src/agent_foundation/utils/observability.py [hunk 2]
```diff
@@ -17,6 +17,7 @@
 import google.auth
 import google.auth.transport.requests
 import grpc
+from fastapi import FastAPI
 from google.auth.exceptions import DefaultCredentialsError
 from google.auth.transport.grpc import AuthMetadataPlugin
 from google.cloud.logging_v2.services.logging_service_v2 import (
```
[the remark this answers is not in the document]
Comment (doughayden) [reply src/agent_foundation/utils/observability.py:23]: false positive, ci passed

File: docs/observability.md [hunk 3]
```diff
@@ -111,6 +111,20 @@ OpenTelemetry resource attributes uniquely identify your service instances in tr
...
 
+### Custom Gen AI Usage Attributes
+
+`LoggingCallbacks.after_model` (registered as the root agent's `after_model_callback`) supplements the genai instrumentor's auto-emitted attributes with three token counts extracted from `GenerateContentResponseUsageMetadata`:
+
+| Attribute | Source field | Semconv status |
+|---|---|---|
+| `gen_ai.usage.cache_read.input_tokens` | `cached_content_token_count` | Canonical (Development stability) — matches "tokens served from provider-managed cache" |
+| `gen_ai.usage.reasoning_tokens` | `thoughts_token_count` | Draft / convention-consistent — name not yet final in semconv, but converging across Gemini / OpenAI / Anthropic telemetry |
+| `gen_ai.usage.tool_use.input_tokens` | `tool_use_prompt_token_count` | Custom — no semconv equivalent yet; name follows the `cache_read.input_tokens` pattern for symmetry |
```
[the remark this answers is not in the document]
Comment (doughayden) [reply docs/observability.md:122]: table is self-descriptive, not changing
